File: 2022/day12/day12.pl
#!/usr/bin/env perl
use v5.36;
use experimental qw(signatures);

my @map = map { chomp; [split ''] } <>;

sub elevation($letter) {
  $letter =~ tr /SE/az/;
  return ord($letter) - ord('a');
}

my (@alternates, %edges, $start, $end);

foreach my $i (0..$#map) {
  my @row = @{$map[$i]};
  foreach my $j (0..$#row) {
    my $letter = $row[$j];
    my $coords = "$i,$j";
    if ($letter eq 'S') {
      $start = $coords;
    } elsif ($letter eq 'E') {
      $end = $coords;
    } elsif ($letter eq 'a') {
      push(@alternates, $coords);
    }
    my $level = elevation($letter);
    foreach my $delta ([-1,0],[0,1],[1,0],[0,-1]) {
      my ($di, $dj) = @$delta;
      my $ni = $i + $di;
      if (0 <= $ni && $ni < @map) {
        my $nj = $j + $dj;
        if (0 <= $nj && $nj < @{$map[$ni]}) {
          my $nlevel = elevation($map[$ni][$nj]);
          push @{$edges{$coords}},"$ni,$nj" if $nlevel <= $level + 1;
        }
      }
    }
  }
}

foreach my $part (0..1) {
  my %heads = ($start => 1);
  if ($part) {
    @heads{@alternates} = (1) x @alternates;
  }

  my $distance = 0;
  while (!$heads{$end}) {
    $distance++;
    my %copy = %heads;
    foreach my $node (keys %copy) {
      foreach my $neighbor (@{$edges{$node}}) {
        $heads{$neighbor} = 1;
      }
    }
  }

  say "Part ${\($part+1)}: $distance";
}

File: 2022/day12/day12.py
#!/usr/bin/env python
from collections import defaultdict
import sys

with (open(sys.argv[1]) if len(sys.argv) > 1 else sys.stdin) as source:
    elevation_map = [line for line in source]

def elevation(letter):
    if letter == 'S':
        letter = 'a'
    elif letter == 'E':
        letter = 'z'
    return ord(letter) - ord('a')

alternates = []
edges = defaultdict(lambda: [])

for i, row in enumerate(elevation_map):
    for j, letter in enumerate(row):
        coords = (i, j)
        if letter == 'S':
            start = coords
        elif letter == 'E':
            end = coords
        elif letter == 'a':
            alternates.append(coords)

        level = elevation(letter)
        for di, dj in ((-1,0),(0,1),(1,0),(0,-1)):
            ni = i + di
            if 0 <= ni  and ni < len(elevation_map):
                nj = j + dj
                if 0 <= nj and nj < len(elevation_map[ni]):
                    nlevel = elevation(elevation_map[ni][nj])
                    if nlevel <= level + 1:
                        edges[coords].append((ni,nj))

for part in range(2):
    heads = set([start])
    if part:
        heads = heads.union(alternates)

    distance = 0;
    while not end in heads:
        distance += 1
        new_heads = heads.copy()
        for node in heads:
            for neighbor in edges[node]:
                new_heads.add(neighbor)
        heads = new_heads

    print(f'Part {part+1}: {distance}')
